File: bindings/python/src/datalevin/vector.py
"""High-level Python wrapper for standalone Datalevin vector indexes."""

from __future__ import annotations

from ._convert import to_python
from ._interop import _BINDINGS
from ._resource import ResourceWrapper


class VectorIndex(ResourceWrapper):
    """KV-backed standalone vector index."""

    def __init__(self, handle) -> None:
        super().__init__(handle, _BINDINGS.close_vector_index, _BINDINGS.vector_index_closed, "vector index")

    def add_vec(self, vec_ref, vec_data) -> "VectorIndex":
        """Add one vector to the index."""

        _BINDINGS.vector_add_vec(self.raw_handle(), vec_ref, vec_data)
        return self

    def remove_vec(self, vec_ref) -> "VectorIndex":
        """Remove all vectors associated with a reference."""

        _BINDINGS.vector_remove_vec(self.raw_handle(), vec_ref)
        return self

    def vec_indexed(self, vec_ref) -> bool:
        """Return whether a vector reference is indexed."""

        return bool(_BINDINGS.vector_indexed(self.raw_handle(), vec_ref))

    def search_vec(self, query_vec, opts=None):
        """Search the vector index."""

        return to_python(_BINDINGS.vector_search(self.raw_handle(), query_vec, opts))

    def re_index(self, opts=None) -> "VectorIndex":
        """Rebuild the vector index."""

        self._handle = _BINDINGS.vector_re_index(self.raw_handle(), opts)
        return self

    def clear(self) -> "VectorIndex":
        """Clear this vector index from memory and disk."""

        handle = self.raw_handle()
        _BINDINGS.vector_clear(handle)
        self._handle = None
        return self

    def force_checkpoint(self) -> "VectorIndex":
        """Force vector checkpoint persistence to the backing KV store."""

        _BINDINGS.vector_force_checkpoint(self.raw_handle())
        return self

    def info(self):
        """Return vector index metadata."""

        return to_python(_BINDINGS.vector_info(self.raw_handle()))

    def checkpoint_state(self):
        """Return checkpoint metadata for this vector index."""

        return to_python(_BINDINGS.vector_checkpoint_state(self.raw_handle()))

File: bindings/javascript/src/vector.js
import util from "node:util";

import { _BINDINGS } from "./interop.js";
import { toJsResult } from "./result.js";

export class VectorIndex {
  constructor(handle) {
    this._handle = handle;
  }

  rawHandle() {
    if (this._handle === null) {
      throw new Error("vector index is closed.");
    }
    return this._handle;
  }

  async closed() {
    if (this._handle === null) {
      return true;
    }
    return Boolean(await _BINDINGS.vectorIndexClosed(this.rawHandle()));
  }

  async close() {
    if (this._handle === null) {
      return;
    }
    const handle = this._handle;
    this._handle = null;
    await _BINDINGS.closeVectorIndex(handle);
  }

  async addVec(vecRef, vecData) {
    await _BINDINGS.vectorAddVec(this.rawHandle(), vecRef, vecData);
    return this;
  }

  async removeVec(vecRef) {
    await _BINDINGS.vectorRemoveVec(this.rawHandle(), vecRef);
    return this;
  }

  async vecIndexed(vecRef) {
    return Boolean(await _BINDINGS.vectorIndexed(this.rawHandle(), vecRef));
  }

  async searchVec(queryVec, opts = null) {
    return toJsResult(await _BINDINGS.vectorSearch(this.rawHandle(), queryVec, opts));
  }

  async reIndex(opts = null) {
    this._handle = await _BINDINGS.vectorReIndex(this.rawHandle(), opts);
    return this;
  }

  async clear() {
    await _BINDINGS.vectorClear(this.rawHandle());
    this._handle = null;
    return this;
  }

  async forceCheckpoint() {
    await _BINDINGS.vectorForceCheckpoint(this.rawHandle());
    return this;
  }

  async info() {
    return toJsResult(await _BINDINGS.vectorInfo(this.rawHandle()));
  }

  async checkpointState() {
    return toJsResult(await _BINDINGS.vectorCheckpointState(this.rawHandle()));
  }

  toString() {
    return `<${this.constructor.name} ${this._handle === null ? "closed" : "open"}>`;
  }

  [util.inspect.custom]() {
    return this.toString();
  }

  async [Symbol.asyncDispose]() {
    await this.close();
  }
}
